teps/0011-redirecting-step-inputs-and-outputs.md

```diff
@@ -0,0 +1,143 @@
+
+```go
+type Step struct {
+    ...
+    // Path to redirect stdin from on container's local filesystem.
+    // +optional
+    Stdin string `json:"stdin,omitempty"`
+    // Path to redirect stdout to on container's local filesystem.
+    // +optional
+    Stdout string `json:"stdout,omitempty"`
+    // Path to redirect stderr to on container's local filesystem.
+    // +optional
+    Stderr string `json:"stderr,omitempty"`
+}
+```
+
+Once `Stdout` or `Stderr` is specified, the corresponding output stream will be duplicated to both the given file and the standard output stream of the container, so users can still view the output through the Pod log API. If both `Stdout` and `Stderr` are set to the same value, the two streams will be written to two file descriptors referring to the same file description, so outputs from both stream will be interleaved in the same file, but there will be no ordering guarantee on the data. The `Stdin` and `Stdout` (or `Stderr`) fields of a step cannot be set to the same value, or the specified file will be truncated and result in an empty stdin.  If multiple steps' `Stdout` fields are set to the same value, the file content will be overwritten by the last outputting step.
+
+Variable substitution will be applied to the new fields, so one could specify `$(results.<name>.path)` to the `Stdout` field to extract the stdout of a step into a Task result. No new variable substitution for accessing the values of `Stdout` and `Stderr` fields will be provided so variable substitution can remain single-pass.
+
+#### Example Usage
+
+Redirecting stdout of `boskosctl` to `jq` and publish the resulting `project-id` as a Task result:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: Task
+metadata:
+  name: boskos-acquire
+spec:
+  results:
+  - name: project-id
+  steps:
+  - name: boskosctl
+    image: gcr.io/k8s-staging-boskos/boskosctl
+    args:
+    - acquire
+    - --server-url=http://boskos.test-pods.svc.cluster.local
+    - --owner-name=christie-test-boskos
+    - --type=gke-project
+    - --state=free
+    - --target-state=busy
+    stdout: /data/boskosctl-stdout
+    volumeMounts:
+    - name: data
+      mountPath: /data
+  - name: parse-project-id
+    image: imega/jq
+    args:
+    - -r
+    - '.name'
+    stdin: /data/boskosctl-stdout
+    stdout: $(results.project-id.path)
+    volumeMounts:
+    - name: data
+      mountPath: /data
+  volumes:
+  - name: data
+```
+
+### Risks and Mitigations
+
+* Users might mistakenly specify paths not shared among steps for redirection. Potentially Tekton could implement a validation to warn users about such misuses.
+
+* If the stdout/stderr of a step is set to the path of a Task result and the step prints too many data, the result manifest would become too large. To prevent this, the `entrypoint` binary should calculate the maximum size allowed for a given result file, and fail the step if the result is too large.
```

Comment:
This would mean my Task fails because it couldn't have written to the result path (too much to write) even though the step did complete successfully ?

Reply:
Failing might be a bit too strong, but the purpose is to surface such problem to users so when a subsequent task tries to consume the result and get an unexpected values, it's easier for users to debug.

We could discuss what's the strategy to handle this problem in another TEP, since this is already an existing problem. I do want to bring it up here because capturing stdout can make it surface more easily.

Reply:
Just for reference, we do already fail a Task if the size of the termination message gets ballooned by results / other data: https://github.com/tektoncd/pipeline/blob/master/cmd/entrypoint/main.go#L86

Reply:
Ah I missed that. Thanks for pointing this out! Then this would not be a problem then.

Reply:
Updating the TEP to reflect that the logic is already in place.